components/datetime/src/metazone.rs

```diff
@@ -0,0 +1,130 @@
+// This file is part of ICU4X. For terms of use, please see the file
+// called LICENSE at the top level of the ICU4X source tree
+// (online at: https://github.com/unicode-org/icu4x/blob/main/LICENSE ).
+
+use crate::provider::time_zones::{MetaZoneId, TimeZoneBcp47Id};
+
+use crate::error::DateTimeFormatError;
+use crate::provider::time_zones::MetaZonePeriodV1Marker;
+use icu_calendar::DateTime;
+use icu_calendar::Iso;
+use icu_locid::Locale;
+use icu_provider::prelude::*;
+use zerovec::ule::AsULE;
+
+pub struct MetaZoneCalculator {
+    pub(super) metazone_period: DataPayload<MetaZonePeriodV1Marker>,
+}
+
+/// [`MetaZoneCalculator`] uses data from the [data provider], the selected [`Locale`] to calculate
+/// metazone id.
+impl MetaZoneCalculator {
+    /// Constructor that loads data before calculating metazone id.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu_locid::locale;
+    /// use icu::datetime::metazone::MetaZoneCalculator;
+    ///
+    /// let provider = icu_testdata::get_provider();
+    /// let mzc = MetaZoneCalculator::new(
+    ///     locale!("en"),
+    ///     &provider,
+    /// );
+    ///
+    /// assert!(mzc.is_ok());
+    /// ```
+    pub fn new<L, ZP>(locale: L, zone_provider: &ZP) -> Result<Self, DateTimeFormatError>
+    where
+        L: Into<Locale>,
+        ZP: ResourceProvider<MetaZonePeriodV1Marker> + ?Sized,
+    {
+        let locale = locale.into();
+        let metazone_period = zone_provider
+            .load_resource(&DataRequest {
+                options: ResourceOptions::from(&locale),
+                metadata: Default::default(),
+            })?
+            .take_payload()?;
+        Ok(Self { metazone_period })
+    }
+
+    /// Calculate metazone id from timezone id and local datetime.
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use icu::datetime::metazone::MetaZoneCalculator;
+    /// use icu_calendar::DateTime;
+    /// use icu_datetime::provider::time_zones::{MetaZoneId, TimeZoneBcp47Id};
+    /// use icu_locid::locale;
+    /// use tinystr::tinystr;
+    ///
+    /// let provider = icu_testdata::get_provider();
+    /// let mzc = MetaZoneCalculator::new(locale!("en"), &provider);
```

Comment:
Suggestion: for the test here, put a `.expect("data exists")` so that below you don't need to `.as_ref().unwrap()` each time